cli/src/merge_tools/builtin.rs

```diff
@@ -245,29 +278,36 @@ pub fn make_diff_files(
 ) -> Result<Vec<scm_record::File<'static>>, BuiltinToolError> {
     let mut files = Vec::new();
     for changed_path in changed_files {
-        let FileInfo {
-            file_mode: left_file_mode,
-            contents: left_contents,
-        } = read_file_contents(store, left_tree, changed_path)?;
-        let FileInfo {
-            file_mode: right_file_mode,
-            contents: right_contents,
-        } = read_file_contents(store, right_tree, changed_path)?;
-
+        let left_info = read_file_contents(store, left_tree, changed_path)?;
+        let right_info = read_file_contents(store, right_tree, changed_path)?;
         let mut sections = Vec::new();
-        if left_file_mode != right_file_mode
-            && left_file_mode != scm_record::FileMode::absent()
-            && right_file_mode != scm_record::FileMode::absent()
-        {
+
+        if should_render_mode_section(&left_info, &right_info) {
             sections.push(scm_record::Section::FileMode {
                 is_checked: false,
-                before: left_file_mode,
-                after: right_file_mode,
+                before: left_info.file_mode,
+                after: right_info.file_mode,
             });
         }
 
-        match (left_contents, right_contents) {
+        match (left_info.contents, right_info.contents) {
             (FileContents::Absent, FileContents::Absent) => {}
+            (
+                FileContents::Absent,
+                FileContents::Text {
+                    contents: _,
+                    hash: _,
+                    num_bytes: 0,
+                },
+            ) => {}
+            (
+                FileContents::Text {
+                    contents: _,
+                    hash: _,
+                    num_bytes: 0,
+                },
+                FileContents::Absent,
+            ) => {}
```

Comment:
nit: maybe this is also a workaround for the upstream bug? add short comment about that?

Reply:
I don't see this as a workaround necessarily. In this context `FileContents::Absent` means that the file doesn't exist. If the file exists on one side and is empty on the other then we'll render a mode change section. This matcher is to avoid also rendering an empty text lines section that just clutters up the UI.

This probably deserves an inline comment though. I'll add one.